ts/hooks/use_wallet.ts

```diff
@@ -0,0 +1,105 @@
+
+    useEffect(() => {
+        setDispatcher(new Dispatcher(dispatch));
+    }, [dispatch]);
+
+    const handleAccountsChange = useCallback(async () => {
+        await asyncDispatcher.fetchAccountInfoAndDispatchToStoreAsync(
+            currentProviderState,
+            dispatcher,
+            currentNetworkId,
+            true,
+        );
+    }, [currentNetworkId, currentProviderState, dispatcher]);
+
+    const handleNetworksChange = useCallback(
+        (newNetworkId: string) => {
+            const networkId = parseInt(newNetworkId, 10);
+
+            if (networkId) {
+                dispatcher.updateNetworkId(networkId);
+            }
+        },
+        [dispatcher],
+    );
+
+    const cleanupCurrentProvider = useCallback(() => {
+        const provider = currentProviderState.provider as any;
+
+        if (provider.off) {
+            if (accountsChangedRef.current) {
+                provider.off(PROVIDER_ACCOUNTS_CHANGED_EVENT, accountsChangedRef.current);
+            }
+
+            if (networkChangedRef.current) {
+                provider.off(PROVIDER_NETWORK_CHANGED_EVENT, networkChangedRef.current);
+                provider.off(PROVIDER_CHAIN_CHANGED_EVENT, networkChangedRef.current);
+            }
+        }
+    }, [currentProviderState.provider]);
+
+    const connectToWallet = useCallback(async () => {
+        cleanupCurrentProvider();
+        const providerState = providerStateFactory.getInitialProviderState(currentNetworkId);
+
+        const networkId = await providerState.web3Wrapper.getNetworkIdAsync();
+        if (networkId !== currentNetworkId) {
+            dispatcher.updateNetworkId(networkId);
+        }
+
+        await asyncDispatcher.fetchAccountInfoAndDispatchToStoreAsync(providerState, dispatcher, networkId, true);
+
+        const newProvider = providerState.provider as any;
+        if (newProvider.on) {
+            accountsChangedRef.current = handleAccountsChange;
+            newProvider.on(PROVIDER_ACCOUNTS_CHANGED_EVENT, accountsChangedRef.current);
+
+            networkChangedRef.current = handleNetworksChange;
+            for (const networkEvent of [PROVIDER_NETWORK_CHANGED_EVENT, PROVIDER_CHAIN_CHANGED_EVENT]) {
+                newProvider.on(networkEvent, networkChangedRef.current);
+            }
+        }
+
+        analytics.track(constants.STAKING.TRACKING.CONNECT_WALLET, {
+            provider: providerState.displayName || 'UnknownWallet',
+        });
+    }, [cleanupCurrentProvider, currentNetworkId, dispatcher, handleAccountsChange, handleNetworksChange]);
+
+    const logoutWallet = useCallback(() => {
+        cleanupCurrentProvider();
+        dispatcher.setAccountStateLoading();
+    }, [cleanupCurrentProvider, dispatcher]);
+
+    return {
+        connectToWallet: useCallback(() => {
+            connectToWallet().catch((err: Error) => {
+                logUtils.warn(`Failed to connect wallet ${err}`);
```

Comment:
Totally unrelated (but reminded me nonetheless) -- do we have a frontend error monitoring service? Sentry/BugSnag? I would like to hook a few of these critical errors up to some sort of reporting mechanism.

Reply:
Not to my knowledge unfortunately 😢  Would love to get something like Sentry in, have been thinking about it myself but wasn't able to anything about it yet.